File: implementations/java-mjs-validator/Harness.java
import java.util.*;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import main.MainClass$;
import java.io.StringWriter;
import java.io.PrintWriter;
import scala.Tuple2;
import java.io.*;
import io.circe.Json;
import java.util.jar.Manifest;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
* The Harness program implements an application that
* simply take input from Bowtie application using standard
* input and output to the standard output.
*
* @author  Sajal Jain
* @version 1.0
* @since   2023-01-10 
*/
class Harness{
	
	public static boolean started = false;
	
	private static final Logger LOGGER = Logger.getLogger( Harness.class.getName() );
	
	private final ObjectMapper objectMapper = new ObjectMapper().configure(
      DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
	
	private static final String NOT_IMPLEMENTED = "This case is not yet implemented.";
	private static final Map<String, String> UNSUPPORTED_CASES = Map.ofEntries(
		Map.entry("escaped pointer ref", NOT_IMPLEMENTED), Map.entry("empty tokens in $ref json-pointer", NOT_IMPLEMENTED),
		Map.entry("maxLength validation", NOT_IMPLEMENTED),Map.entry("minLength validation", NOT_IMPLEMENTED),
		Map.entry("$id inside an unknown keyword is not a real identifier", NOT_IMPLEMENTED), Map.entry("schema that uses custom metaschema with with no validation vocabulary", NOT_IMPLEMENTED),
		Map.entry("small multiple of large integer", NOT_IMPLEMENTED), Map.entry("$ref to $ref finds detached $anchor", NOT_IMPLEMENTED),
		Map.entry("$ref to $dynamicRef finds detached $dynamicAnchor", NOT_IMPLEMENTED));
	
	public static void main(String[] args) {
		Scanner input = new Scanner(System.in);
		while (true) {
			String line = input.nextLine();
			String output = new Harness().operate(line);
			System.out.println(output);
		}
	}
	
	public String operate(String line){
		String error = "";
		try{
			JsonNode node = objectMapper.readTree(line);
			String cmd = node.get("cmd").asText();
			switch(cmd) {
			  case "start":
				StartRequest startRequest = objectMapper.treeToValue(node, StartRequest.class);
				long version = startRequest.version();
				if(version == 1){
					InputStream is = getClass().getResourceAsStream("META-INF/MANIFEST.MF");
					var attributes = new Manifest(is).getMainAttributes();
					started = true;
					JSONObject message = new JSONObject();
					message.put("ready", true);
					message.put("version", 1);
					JSONObject implementation = new JSONObject();
					implementation.put("language", "scala");
					implementation.put("name", attributes.getValue("Implementation-Name"));
					implementation.put("version", attributes.getValue("Implementation-Version"));
					implementation.put("homepage", "https://gitlab.lip6.fr/jsonschema/modernjsonschemavalidator");
					implementation.put("issues", "https://gitlab.lip6.fr/jsonschema/modernjsonschemavalidator/issues");
					JSONArray dialects = new JSONArray();
					dialects.add("https://json-schema.org/draft/2020-12/schema");
					implementation.put("dialects", dialects);
					message.put("implementation", implementation);
					return message.toJSONString();
				}
				break;
			  case "dialect":
				if(started != true){
					throw new RuntimeException("Bowtie hasn't started!");
				}
				DialectRequest dialectRequest = objectMapper.treeToValue(node, DialectRequest.class);
				return "{ \"ok\" : false }";
			  case "run":
				if(started != true){
					throw new RuntimeException("Bowtie hasn't started!");
				}
				RunRequest runRequest = objectMapper.treeToValue(node, RunRequest.class);
				
				try{
					if(UNSUPPORTED_CASES.containsKey(runRequest.testCase().description())){
						return skipMsg(UNSUPPORTED_CASES.get(runRequest.testCase().description()), runRequest.seq().asLong());
					}
					
					JSONArray resultArray = new JSONArray();
					
					for (Test test : runRequest.testCase().tests()) {
						String instance = test.instance().toString();
						MainClass$ m = MainClass$.MODULE$;
						Tuple2<Object, Json> results = m.validateInstance(runRequest.testCase().schema().toString(), instance);
						
						JSONObject result = new JSONObject();
						result.put("valid", (boolean) results._1());
						resultArray.add(result);
					}
					
					JSONObject out = new JSONObject();
					out.put("seq", runRequest.seq().asLong());
					out.put("results", resultArray);
					return out.toJSONString();
				}
				catch(Exception e){
					String msg = getDetailedMessage(e, runRequest.testCase().schema().toString());
					error = errorMsg(msg, runRequest.seq().asLong());
					return error;
				}
			  case "stop":
				if(started != true){
					throw new RuntimeException("Bowtie hasn't started!");
				}
				System.exit(0);
			}
		}catch(Exception e){
			error = errorMsg(e.getMessage(), -1);
			return error;
		}
		error = errorMsg("Send correct command", -1);
		return error;
	}
	
	public static String errorMsg(String message, long seq){
		JSONObject traceBack = new JSONObject(); 
		traceBack.put("traceBack", message);
		JSONObject error = new JSONObject(); 
		error.put("errored", true);
		error.put("seq", seq);
		error.put("context",traceBack);
		return error.toJSONString();
	}
	
	public static String skipMsg(String message, long seq){
		JSONObject error = new JSONObject();
		error.put("skipped", true);
		error.put("seq", seq);
		error.put("message",message);
		return error.toJSONString();
	}
	
	public static String getDetailedMessage(Exception e, String schema){
		StringWriter sw = new StringWriter();
		e.printStackTrace(new PrintWriter(sw));
		return sw.toString() + " " + schema;
	}
}

record StartRequest(int version) {}

record DialectRequest(String dialect) {}

record RunRequest(JsonNode seq, @JsonProperty("case") TestCase testCase) {}

record TestCase(String description, String comment, JsonNode schema, JsonNode registry, List<Test> tests) {}

record Test(String description, String comment, JsonNode instance, boolean valid) {}

File: implementations/scala-rc-circe-json-validator/Harness.scala
import io.circe.generic.auto._
import io.circe.parser.parse
import io.circe.syntax.EncoderOps
import io.circe.{Json, Decoder, ParsingFailure}
import java.io.InputStream
import java.util.Scanner
import java.util.jar.Manifest
import net.reactivecore.cjs.{Loader, ResolveFailure, Result}
import net.reactivecore.cjs.resolver.Downloader

class BowtieRcCirceJsonValidator {
  def operate(line: String) = {
    val node: Json = parse(line) match {
      case Right(json)          => json
      case Left(parsingFailure) => throw parsingFailure
    }

    val cmd: String = (node \\ "cmd").headOption match {
      case Some(value) => value.asString.getOrElse("")
      case None        => throw new RuntimeException("Failed to get cmd")
    }
    cmd match {
      case "start"   => println(start(node))
      case "dialect" => println(dialect(node))
      case "run"     => println(run(node))
      case "stop"    => System.exit(0)
      case default   => throw new IllegalArgumentException(s"Unknown command: $default")
    }
  }

  def start(node: Json): String = {
    BowtieRcCirceJsonValidator.started = true
    val startRequest: StartRequest = decodeTo[StartRequest](node)
    if (startRequest.version != 1) {
      throw new IllegalArgumentException(s"Unsupported IHOP version: ${startRequest.version}")
    }

    val is: InputStream = getClass.getResourceAsStream("META-INF/MANIFEST.MF")
    val attributes = new Manifest(is).getMainAttributes

    val implementation: Implementation = Implementation(
      "scala",
      attributes.getValue("Implementation-Name"),
      attributes.getValue("Implementation-Version"),
      List("https://json-schema.org/draft/2020-12/schema"),
      "https://github.com/reactivecore/rc-circe-json-schema",
      "https://github.com/reactivecore/rc-circe-json-schema/issues",
      "https://github.com/reactivecore/rc-circe-json-schema",
      System.getProperty("os.name"),
      System.getProperty("os.version"),
      Runtime.version().toString,
      List()
    )
    StartResponse(startRequest.version, implementation).asJson.noSpaces
  }

  def dialect(node: Json): String = {
    if (!BowtieRcCirceJsonValidator.started) {
      throw new IllegalArgumentException("Not started!")
    }
    DialectResponse(true).asJson.noSpaces
  }

  def run(node: Json): String = {
    if (!BowtieRcCirceJsonValidator.started) {
      throw new IllegalArgumentException("Not started!")
    }

    val runRequest: RunRequest = decodeTo[RunRequest](node)
    val testCase = runRequest.testCase
    try {
      val registry: Map[String, Json] = testCase.registry.getOrElse(Map.empty)
      val downloader = new InMemoryDownloader(registry)
      val validator = Loader(downloader).fromJson(testCase.schema).right.get

      var resultArray = Vector.empty[Json]
      runRequest.testCase.tests.foreach { test =>
        val instance: String = test.instance.noSpaces
        val result = Json.obj("valid" -> validator.validate(test.instance).isSuccess.asJson)
        resultArray :+= result
      }
      RunResponse(runRequest.seq, resultArray).asJson.noSpaces
    } catch {
      case e: Exception =>
        val error: ErrorContext = ErrorContext(e.getMessage(), e.getStackTrace().mkString("\n"))
        RunErroredResponse(runRequest.seq, context = error).asJson.noSpaces
    }
  }

  def decodeTo[Request: Decoder](json: Json): Request = {
    json.as[Request] match {
      case Right(value)          => value
      case Left(decodingFailure) => throw decodingFailure
    }
  }
}


class InMemoryDownloader(registry: Map[String, Json]) extends Downloader[Result] {
  override def loadJson(url: String): Result[Json] = {
    registry.get(url) match {
      case Some(json) => Right(json)
      case None => Left(ResolveFailure(s"Failed to resolve URL: $url"))
    }
  }
}

case class Implementation(language: String, name: String, version: String,
                          dialects: List[String], homepage: String, issues: String,
                          source: String, os: String, os_version: String,
                          language_version: String, links: List[Link])

case class Link(url: String, description: String)

case class StartRequest(version: Int)

case class StartResponse(version: Int, implementation: Implementation)

case class DialectRequest(dialect: String)

case class DialectResponse(ok: Boolean)

case class RunRequest(seq: Json, testCase: TestCase)
implicit val decodeRunRequest: Decoder[RunRequest] = Decoder.forProduct2("seq", "case")(RunRequest.apply)

object RunRequest {}

case class RunResponse(seq: Json, results: Vector[Json])

case class RunSkippedResponse(seq: Json, skipped: Boolean = true, message: Option[String] = None)

case class RunErroredResponse(seq: Json, errored: Boolean = true, context: ErrorContext)

case class ErrorContext(message: String, traceback: String)

case class TestCase(description: String, comment: Option[String], schema: Json, registry: Option[Map[String, Json]], tests: List[Test])

case class Test(description: String, comment: Option[String], instance: Json, valid: Option[Boolean])

case class SkippedTest(skipped: Boolean = true, message: Option[String] = None)

case class TestSkip(description: String, message: String)

object BowtieRcCirceJsonValidator {
  var started: Boolean = false

  def main(args: Array[String]): Unit = {
    val input = new Scanner(System.in)
    while (true) {
      new BowtieRcCirceJsonValidator().operate(input.nextLine())
    }
  }
}
